\documentclass{bioinfo}
\usepackage{amsmath}
\usepackage{amssymb}

\DeclareMathOperator*{\argmax}{arg\,max}
\newcommand{\s}{\mathbf{\Sigma}}

\newcommand{\citeeigen}{(http://eigen.tuxfamily.org)}

\copyrightyear{2016} \pubyear{2016}

\newcommand{\figwidthp}{0.43}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Applications Note}

\input{numbers.tex}

\begin{document}
\firstpage{1}

\subtitle{Genetics and population analysis}

\title[Fast SCCA]{FlashPCA: fast sparse canonical correlation analysis of genomic data}
\author[Sample \textit{et~al}.]{Gad Abraham\,$^{\text{\sfb 1,2}*}$
and Michael Inouye\,$^{\text{\sfb 1,2}}$}
\address{$^{\text{\sf 1}}$ Centre for Systems Genomics, School of
BioSciences, University of Melbourne, Parkville 3010, VIC, Australia. \\
$^{\text{\sf 2}}$ Department of Pathology, Faculty of Medicine, Dentistry, and
Health Sciences, University of Melbourne,\\ Parkville 3010, VIC, Australia.}

\corresp{$^\ast$To whom correspondence should be addressed.}
\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}
\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Summary:} Sparse canonical correlation analysis (SCCA) is a
useful approach for correlating one set of measurements, such as single
nucleotide polymorphisms (SNPs), with another set of measurements, such as gene
expression levels.  We present a fast implementation of SCCA, enabling rapid
analysis of hundreds of thousands of SNPs together with thousands of phenotypes.
Our approach is implemented both as an R package \texttt{flashpcaR} and within
the standalone commandline tool \texttt{flashpca}.\\
\textbf{Availability and implementation:}
\href{https://github.com/gabraham/flashpca}{https://github.com/gabraham/flashpca}
\\ \textbf{Contact:}
\href{gad.abraham@unimelb.edu.au}{gad.abraham@unimelb.edu.au}\\
\textbf{Supplementary information:} Supplementary data are available at
\textit{Bioinformatics} online.}

\maketitle

\input{text.tex}

\bibliographystyle{natbib}
\bibliography{paper}

\end{document}

