\documentclass[a4paper,12pt]{article}


%\usepackage{a4wide}
\usepackage{geometry}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage[numbers]{natbib}

\geometry{
   includeheadfoot,
   margin=2.54cm
}


\hypersetup{
   pdftitle={},
   pdfauthor={Gad Abraham},
   colorlinks=true,
   citecolor=black,
   filecolor=black,
   linkcolor=black,
   urlcolor=black
}

\input{numbers.tex}

\author{Gad Abraham and Michael Inouye}

\title{Fast sparse canonical correlation with flashpca --- Supplementary
Material}

\begin{document}

\maketitle

\section{Reproducibility}

Code to reproduce these experiments is at
\url{https://github.com/gabraham/scca-paper}.

\section{HapMap data preprocessing and quality control}

The HapMap3 phase III~\citep{hapmap2010} genotypes were obtained from
\url{ftp://ftp.ncbi.nlm.nih.gov/hapmap/genotypes/2009-01_phaseIII/plink_format/}.
Gene expression levels were obtained from
\url{http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-264}.

We excluded individuals who were non-founders, had genotyping
missingness~${>}1\%$, or did not have matching gene expression data,
resulting in~\nindiv individuals. We excluded non-autosomal SNPs, SNPs
with MAF~${<}5\%$, missingness~${>}1\%$, and deviation from Hardy-Weinberg
equilibrium~$P{<}5\times10^{-6}$ using PLINK~1.9~\citep{purcell2007,Chang2015},
leaving~\nsnps autosomal SNPs (see below). The remaining missing genotypes were
randomly imputed according to the frequencies of the non-missing observations.

For the gene expression data, we used a subset consisting of the~21,800 probes
that were analysed by~\citep{Stranger2012}, utilising the original authors'
normalised data. Following~\citep{Stranger2012}, we performed PCA on the
genotypes within each population, and for the GIH, MEX, MKK, and LWK regressed
out~10 PCs of the genotypes (as well as intercept) from the corresponding gene
expression levels, in order to adjust for the higher levels of admixture within
these populations.  We further filtered probes with low variance
(std.~dev.~${<}0.1$), leaving~\ngenes probes. Both the gene expression levels and
the genotypes were standardised to zero-mean and unit-variance.

\section{Comparison of predictive power in simulation}

Utilising 10,000 SNPs from chromosome~1, we simulated 1,000 gene expression
levels as
$$
\mathbf{Y} = \mathbf{X} \mathbf{B} + \mathbf{E},
$$
where $\mathbf{X}$ are the genotypes ($n \times p$ matrix), $\mathbf{B}$
is a $p \times m$ matrix of weights, and $\mathbf{E}$ is an $n \times m$
matrix representing the error (noise). To match the sparsity assumptions of
SCCA, $\mathbf{B}$ was chosen to be a mixture of weights $\{0.001, 1\}$ with
proportions 0.9999 and 0.0001 (across all $n \times m$ entries), respectively,
(0.001 rather than zero was used to prevent some probes from having zero
genetic variance). Each column $k=1,\hdots,m$ of $\mathbf{E}$ was $E_k \sim
\mathcal{N}(0, \frac{1-h^2}{h^2} \mbox{var}((\mathbf{X}\mathbf{B})_k))$,
and $h^2=0.1$.

We used 3-fold cross-validation to compare \texttt{flashpcaR::scca} and
\texttt{PMA::CCA}, over a~2D grid of $30\times25$ penalties, estimating one
pair of canonical vectors. The final predictive power was computed as the
average Pearson correlation $\bar{\rho}$ in the~$k=1,\hdots,3$ test folds:
$$
\bar{\rho} = \frac{1}{3} \sum_{k=1}^3
   \mbox{Cor}(\mathbf{X}_{test}^k u^k, \mathbf{Y}_{test}^k v^k).
$$

\section{Timing experiments}

For timing of \texttt{flashpcaR::scca} and \texttt{PMA::CCA}, we used contiguous
subsets of chromosome~1 (1000, 5000, 10,000, 20,000, and 50,000 SNPs, out
of~\ngenes SNPs in total) and contigous subsets of the~\ngenes gene expression
probes (1000, 10,000, and all~\ngenes probes).

We used the \textsf{R} package \texttt{microbenchmark}~\citep{Mersmann2015}
to run~30 replications of each timing experiment.  For all experiments we
estimated one pair of canonical vectors $(u_1, v_1)$.  For the results in the
main text, we initialised (``warm started'') $v_1$ to a standard normally-
distributed vector of variates $\sim\mathcal{N}(0,1)$. \texttt{PMA::CCA} and
\texttt{flashpcaR::scca} (but not the commandline version \texttt{flashpca})
allow the user to provide their own initialisation, and we experimented with
other forms, including using the column means of the gene expression data
and the rank-1 singular value decomposition (SVD) $\mathbf{X}^T \mathbf{Y}
\approx u_1 d_1 v_1^T$) using \texttt{flashpcaR::flashpca}. The overall
trend of \texttt{flashpca} being several-fold faster than \texttt{PMA}
was consistent across all three initialistion methods~(Figure~\ref{fig:s01}).

\begin{figure}[!tpb]
\centering
\includegraphics[width=\textwidth]{scca_timing_full-crop.pdf}
\caption{
Timing (median of 30 runs) of SCCA implemented in (i) the \texttt{flashpcaR}
(\textsf{R} package) and (ii) \texttt{flashpca} (stand-alone commandline tool),
compared with SCCA from \texttt{PMA}, using subsets of the HapMap3 dataset with gene
expression levels as phenotypes. We compared three schemes for initialising
$v_1$: (i) ``mean'': column means of the gene expression data; (ii) ``rand'':
normally-distributed variates $\mathcal{N}(0, 1)$; and (iii) ``svd'': 1st
right singular value of $\mathbf{X}^T \mathbf{Y}$.
}
\label{fig:s01}
\end{figure}

All experiments were run in \textsf{R}~3.2.2~\citep{R} (with the original
LAPACK and BLAS libraries included in \textsf{R}) on 64-bit Ubuntu Linux~12.04
on an Intel Xeon CPU~E7-4830 v2 @~2.20GHz. Time for the commandline
\texttt{flashpca} include loading of data into RAM. We used flashpca~v1.2.6
(\url{https://github.com/gabraham/flashpca}) and PMA~v1.0.9~\citep{Witten2013}.
For \texttt{PMA::CCA}, we increased the maximum number of iterations to match
that used by \texttt{flashpcaR::scca} (default=1000), in order to prevent
early termination of the algorithm before adequate numerical convergence was
achieved. 

\bibliographystyle{unsrtnat}
\bibliography{paper}

\end{document}

